README
NAME
    WWW::GoodData - Client library for GoodData REST-ful API

SYNOPSIS
      use WWW::GoodData;
      my $gdc = new WWW::GoodData;
      print $gdc->get_uri('md', { title => 'My Project' });

DESCRIPTION
    WWW::GoodData is the client for GoodData JSON-based API built atop
    WWW::GoodData::Agent client agent, with focus on usefullness and
    correctness of implementation.

    It provides code for navigating the REST-ful API structure as well as
    wrapper funcitons for common actions.

METHODS
    new [PARAMS]
        Create a new client instance.

        You can optionally pass a hash reference with properties that would
        be blessed, otherwise a new one is created. Possible properties
        include:

        agent   A WWW::GoodData::Agent instance to use.

        auto_destroy
                Sets automatic logout in destroying object.

                Defaults to 1.

        retries A number of retries to obtain results of asynchronous tasks,
                such as report exports or data uploads. See poll.

                Defaults to 3600 (delay of one hour).

        verbose Adds handler for printing of request information to agent
                When versbose mode is set. Password in 'postUserLogin'
                parameter will be hidden.

                Examples of verbose output: GET
                https://secure.gooddata.com/gdc POST
                https://secure.gooddata.com/gdc/account/login
                {"postUserLogin":{"login":"bear@gooddata.com","verify_level"
                :2,"remember":0,"password":"***"}}

                Defaults to 0.

    links PATH
        Traverse the links in resource hierarchy following given PATH,
        starting from API root ("gdc" by default).

        PATH is an array of dictionaries, where each key-value pair matches
        properties of a link. If a plain string is specified, it is
        considered to be a match against category property:

          $gdc->links('md', { 'category' => 'projects' });

        The above call returns a list of all projects, with links to their
        metadata resources.

    get_uri PATH
        Follows the same samentics as links() call, but returns an URI of
        the first matching resource instead of complete link structure.

    login EMAIL PASSWORD
        Obtain a SST (login token).

    logout
        Make server invalidate the client session and drop credential
        tokens.

        Is called upon destruction of the GoodData client instance.

    change_passwd OLD NEW
        Change user password given the old and new password.

    projects
        Return array of links to project resources on metadata server.

    delete_project IDENTIFIER
        Delete a project given its identifier.

    create_project TITLE SUMMARY TEMPLATE DRIVER TOKEN
        Create a project given its title and optionally summary, project
        template, db engine driver and authorization token return its
        identifier.

        The list of valid project templates is available from the template
        server: <https://secure.gooddata.com/projectTemplates/>.

        Valid db engine drivers are 'Pg' (default) and 'vertica'.

    create_user DOMAIN EMAIL LOGIN PASSWORD FIRST_NAME LAST_NAME PHONE
    COMPANY SSO_PROVIDER
        Create a user given its email, login, password, first name, surname,
        phone and optionally company, sso provider in domain.

        Returns user identifier (URI).

    get_roles PROJECT
        Gets project roles.

        Return array of project roles.

    reports PROJECT
        Return array of links to repoort resources on metadata server.

    compute_report REPORT
        Trigger a report computation and return the URI of the result
        resource.

    export_report REPORT FORMAT
        Submit an exporter task for a computed report (see compute_report),
        wait for completion and return raw data in desired format.

    ldm_picture PROJECT
        Return picture of Logical Data Model (LDM) in PNG format.

    ldm_manage PROJECT MAQL
        Execute MAQL statement for a project.

    upload PROJECT MANIFEST
        Upload and integrate a new data load via Single Loading Interface
        (SLI).

    uploads FILE_PATH FILE UPLOAD_PATH
        Uploads to WebDAV service.

    poll TASK_URL
        Periodically checks status of task page and check
        'wTaskStatus':'status' for 'RUNNING'.

        Returns undef is status is OK.

    create_object_with_expression PROJECT URI TYPE TITLE SUMMARY EXPRESSION
        Create a new metadata object of type TYPE with EXPRESSION as the
        only content.

    create_report_definition PROJECT URI TITLE SUMMARY METRICS DIM FILTERS
        Create a new reportDefinition in metadata.

    DESTROY
        Log out the session with logout unless not logged in.

    dd_pull PROJECT ZIP_FILE DATASETS
        Integrate a custom date dimension load via Single Loading Interface
        (SLI).

SEE ALSO
    *   <http://developer.gooddata.com/api/> -- API documentation

    *   <https://secure.gooddata.com/gdc/> -- Browsable GoodData API

    *   WWW::GoodData::Agent -- GoodData API-aware user agent

CAVEATS
    *       uploads()

            Don't check, when collection exists.

LICENSE AND COPYRIGHT
    Copyright 2017 Michal Josef Spacek

    Copyright 2011, 2012, 2013 Lubomir Rintel

    Copyright 2012, 2013 Adam Stulpa, Jan Orel, Tomas Janousek

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

AUTHORS
    Tomas Janousek "tomi@nomi.cz"

    Jan Orel "jan.orel@gooddata.com"

    Lubomir Rintel "lkundrak@v3.sk"

    Michal Josef Spacek "michal.spacek@gooddata.com"

    Adam Stulpa "adam.stulpa@gooddata.com"



Changes
1.12
=======
 - Add 'auto_destroy' constructor parameter and check it in DESTROY().
 - Add dd_pull() method.
 - Add environment support to create_project() method.
 - Add LICENSE file.
 - Add META.json and META.yml files.
 - Add minimal version of Perl.
 - Add README file.
 - Add uploads() method.
 - Add verbose mode.
 - Fix database driver in doc of create_project.
 - Fix missing use strict and use warning.
 - Move maintance of module to Michal Josef Spacek.
 - Move poll() method to private _poll and create new implementation of poll()
   with standardy async polling.
 - Reindent code.
 - Rewrite COPYRIGHT section in doc to more descriptive
   'LICENSE AND COPYRIGHT'.
 - Rewrite tests to recursive way, dir per module.
 - Switch to etl/pull2.


lib/WWW/GoodData/Agent.pm
package WWW::GoodData::Agent;

=head1 NAME

WWW::GoodData::Agent - HTTP client for GoodData JSON-based API

=head1 SYNOPSIS

  use WWW::GoodData::Agent;
  my $ua = new WWW::GoodData::Agent;
  my $metadata = $ua->get ('/md');

=head1 DESCRIPTION

B<WWW::GoodData::Agent> is HTTP user agent that makes it easy for follow
specifics of the GoodData service API, transparently handles conversion
to and from JSON content type and recognizes and handles various kinds
of exceptions and error states.

It is a subclass of L<LWP::UserAgent> and follows its semantics unless
documented otherwise.

=cut

use strict;
use warnings;

use base qw/LWP::UserAgent/;

use JSON;

our $VERSION = '1.12';

=head1 PROPERTIES

=over 4

=item root

L<URI> object pointing to root of the service API.

This is used to resolve relative request paths.

=back

=head1 METHODS

=over 4

=item new ROOT, PARAMS

Creates a new agent instance. First argument is root of
the service API, the rest is passed to L<LWP::UserAgent> as is.

Compared to stock L<LWP::UserAgent>, it has a memory-backed
cookie storage and sets the B<Accept> header to prefer JSON content.

=cut

sub new {
	my ($self, $root, @args) = @_;
	$self = $self->SUPER::new(@args);
	$self->{root} = $root;
	$self->agent("perl-WWW-GoodData/$VERSION ");
	# Not backed by a file yet
	$self->cookie_jar({});
	# Prefer JSON, but deal with whatever else comes in, instead of letting backend return 406s
	$self->default_header(
		Accept => 'application/json;q=0.9, text/plain;q=0.2, */*;q=0.1',
	);
	return $self;
}

=item post URI, BODY, PARAMS

Constructs and issues a POST request.

Compared to stock L<LWP::UserAgent>, the extra body parameter
is encoded into JSON and set as request content, which is the
only way to set the request content.

The rest of parameters are passed to L<LWP::UserAgent> untouched.

=cut

sub post {
	my ($self, $uri, $body, @args) = @_;
	push @args, 'Content-Type' => 'application/json',
		Content => encode_json($body);
	return $self->SUPER::post($uri, @args);
}

=item put URI, BODY, PARAMS

Constructs and issues a PUT request.

Compared to stock L<LWP::UserAgent>, the extra body parameter
is encoded into JSON and set as request content, which is the
only way to set the request content.

The rest of parameters are passed to L<LWP::UserAgent> untouched.

=cut

sub put {
	my ($self, $uri, $body, @args) = @_;
	push @args, 'Content-Type' => 'application/json',
		Content => encode_json($body);
	return $self->SUPER::put($uri, @args);
}

=item delete URI

Convenience method for constructing and issuing a DELETE request.

=cut

sub delete {
	my ($self, $uri) = @_;

	return $self->request(HTTP::Request->new(DELETE => $uri, \@_));
}

=item request PARAMS

This call is common for all request types.

While API is same as stock L<LWP::UserAgent>, relative URIs
are permitted and extra content processing is done with the response.

Namely, errors are either handled or turned into exceptions
and known content types (JSON) are decoded.

=cut

sub request {
	my ($self, $request, @args) = @_;

	# Ignore Basic authentication for our requests
	no strict 'refs';
	local %{'LWP::Authen::Basic::'} = (dummy => 'stash');

	# URI relative to root
	$request->uri($request->uri->abs($self->{root}));

	# Level 2 authentication
	$request->header('X-GDC-AuthTT' => $self->{GDCAuthTT})
		if $self->{GDCAuthTT};

	# Issue the request
	my $response = $self->SUPER::request($request, @args);

	# Pass processed response from subrequest (redirect)
	return $response if not defined $response or ref $response eq 'HASH';

	# Do not bother checking content and type if there's none
	return undef if $response->code == 204;

	# Decode
	my $decoded = eval { decode_json($response->content) }
		if $response->header('Content-Type') =~ /^application\/json(;.*)?/;
	$decoded = {
		type => $response->header('Content-Type'),
		raw => $response->content,
	} unless $decoded;

	# Error handling
	unless ($response->is_success) {
		# Apache::Error exceptions lack error wrapper
		$decoded = $decoded->{error} if exists $decoded->{error};
		my $request_id = $response->header('X-GDC-Request') || "";
		$request_id = " (Request ID: $request_id)" if $request_id;
		die $response->status_line.$request_id unless exists $decoded->{message};
		die sprintf($decoded->{message}, @{$decoded->{parameters}}).$request_id;
	}

	return $decoded;
}

=back

=head1 SEE ALSO

=over

=item *

L<https://secure.gooddata.com/gdc/> -- Browsable GoodData API

=item *

L<LWP::UserAgent> -- Perl HTTP client
        
=back

=head1 LICENSE AND COPYRIGHT

Copyright 2017 Michal Josef Spacek

Copyright 2011, 2012, 2013 Lubomir Rintel

Copyright 2012 Jan Orel

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=head1 AUTHOR

Michal Josef Spacek <michal.spacek@gooddata.com>

Lubomir Rintel C<lkundrak@v3.sk>

=cut

1;


lib/LWP/Authen/Gooddata.pm
package LWP::Authen::Gooddata;

use strict;
use warnings;

our $VERSION = '1.12';

=head1 NAME

LWP::Authen::Gooddata - Handle GoodData HTTP authentication mechanism

=head1 SYNOPSIS

  use WWW::GoodData::Agent;
  my $agent = WWW::GoodData::Agent->new('https://secure.gooddata.com/gdc');
  $agent->post ('/gdc/account/login', ...);
  # The authentication cookie gets obtained transparently here
  $agent->get('/gdc/md');

=head1 DESCRIPTION

B<LWP::Authen::Gooddata> gets loaded and invoked by a L<LWP::UserAgent>,
or its subclass such as L<WWW::GoodData::Agent>, upon reciept of 401
"Unauthorized" response from the server which indicates use of "GoodData"
authentization mechanism in the "WWW-Authenticate" header.

If the challenge indicates the temporary authentization cookie needs
to be refreshed it does so transparently and reissues the request,
otherwise dies with appropriate explanation.

=head1 METHODS

=over 4

=item B<authenticate> [PARAMS]

Called by LWP::UserAgent internally.

=cut

sub authenticate {
	my ($class, $agent, $proxy, $challenge, $response,
		$request, $arg, $size) = @_;

	# Not an useful 401 from backend, pass the response
	return $response unless $challenge->{cookie};

	# Need to obtain super-secure token
	die 'Login required' if $challenge->{cookie} eq 'GDCAuthSST';

	# Everything else should be a token refresh
	die 'Required authentication not supported by client'
		if $challenge->{cookie} ne 'GDCAuthTT';

	# Get the token resource location
	my $token_uri = $challenge->{location};
	# Compat
	unless ($token_uri) {
		$token_uri = $request->uri->clone;
		$token_uri->path ($challenge->{location} or '/gdc/account/token');
		$token_uri->fragment (undef);
	}

	# Refresh the token
	$agent->{GDCAuthTT} = $agent->get(URI->new($token_uri),
		'X-GDC-AuthSST' => $agent->{GDCAuthSST})->{userToken}{token};

	$agent->default_header('X-GDC-AuthTT' => $agent->{GDCAuthTT});
	$request->header(Cookie => '');

	# Retry the request
	return $agent->simple_request ($request);
}

=back

=head1 SEE ALSO

=over

=item *

L<http://developer.gooddata.com/api/auth.html> -- Specification of the GoodData authentization mechanism

=item *

L<LWP::UserAgent> -- The Perl HTTP agent

=back

=head1 BUGS

GoodData authentization mechanism is not an internet standard and thus
puts an interoperability barrier. Unfortunatelly, no standard and widely
supported mechanism provides comparable benefits (mostly server-side
performance coupled with sanity of implementation). Probably an alternative
mechanism should be provided (Basic or Digest, which are both widely
available) for the client to negotiate.

=head1 LICENSE AND COPYRIGHT

Copyright 2017 Michal Josef Spacek

Copyright 2011, 2012, 2013 Lubomir Rintel

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=head1 AUTHOR

Lubomir Rintel C<lkundrak@v3.sk>

=cut

1;


lib/WWW/GoodData.pm
package WWW::GoodData;

=head1 NAME

WWW::GoodData - Client library for GoodData REST-ful API

=head1 SYNOPSIS

  use WWW::GoodData;
  my $gdc = new WWW::GoodData;
  print $gdc->get_uri('md', { title => 'My Project' });

=head1 DESCRIPTION

B<WWW::GoodData> is the client for GoodData JSON-based API
built atop L<WWW::GoodData::Agent> client agent, with focus
on usefullness and correctness of implementation.

It  provides code for navigating the REST-ful API structure as well as
wrapper funcitons for common actions.

=cut

use strict;
use warnings;

use English;
use WWW::GoodData::Agent;
use JSON;
use URI;

our $VERSION = '1.12';
our $root = URI->new('https://secure.gooddata.com/gdc');

=head1 METHODS

=over 4

=item B<new> [PARAMS]

Create a new client instance.

You can optionally pass a hash reference with properties that would be
blessed, otherwise a new one is created. Possible properties include:

=over 8

=item B<agent>

A L<WWW::GoodData::Agent> instance to use.

=item B<auto_destroy>

Sets automatic logout in destroying object.

Defaults to 1.

=item B<retries>

A number of retries to obtain results of asynchronous tasks, such as
report exports or data uploads. See B<poll>.

Defaults to 3600 (delay of one hour).

=item B<verbose>

Adds handler for printing of request information to agent When versbose mode is set.
Password in 'postUserLogin' parameter will be hidden.

Examples of verbose output:
 GET https://secure.gooddata.com/gdc
 POST https://secure.gooddata.com/gdc/account/login {"postUserLogin":{"login":"bear@gooddata.com","verify_level":2,"remember":0,"password":"***"}}

Defaults to 0.

=back

=cut

sub new {
	my $class = shift;
	my $self = shift || {};
	bless $self, $class;
	$self->{agent} ||= WWW::GoodData::Agent->new($root);
	$self->{auto_destroy} //= 1;
	$self->{retries} ||= 3600;
	$self->{verbose} ||= 0;

	$self->{agent}->add_handler('request_prepare', sub {
		my ($request, $ua, $h) = @_;
		my $output = $request->method.' '.$request->uri_canonical;
		my $content = $request->content;
		if ($request->method eq 'POST' && defined $content && $content ne '') {
			$content = $self->_hide_password($content);
			$output .= ' '.$content;
		}
		print "$output\n";
		return;
	}) if $self->{verbose};

	return $self;
}

# API hierarchy traversal Cache
our %links;
sub get_canonical_links {
	my $self = shift;
	my $root = shift;
	my @path = map { ref $_ ? $_ : { category => $_ } } @_;
	my $link = shift @path;

	unless ($links{$root}) {
		my $response = $self->{agent}->get($root);
		# Various ways to get the links
		if (exists $response->{about}) {
			# Ordinary structure with about section
			$links{$root} = $response->{about}{links};
		} elsif (exists $response->{query} and exists $response->{query}{entries}) {
			# Inconsistent query entries
			$links{$root} = $response->{query}{entries};
		} elsif (scalar keys %$response == 1) {
			my @elements = ($response);
			my ($structure) = keys %$response;

			# Aggregated resources (/gdc/account/profile/666/projects)
			@elements = @{$response->{$structure}}
				if ref $response->{$structure} eq 'ARRAY';

			$links{$root} = [];
			foreach my $element (@elements) {
				my $root = $root;
				my ($type) = keys %$element;

				# Metadata with interesting information outside "links"
				if (exists $element->{$type}{links}{self}
					and exists $element->{$type}{meta}) {
					my $link = URI->new($element->{$type}{links}{self})->abs($root);
					push @{$links{$root}}, {
						%{$element->{$type}{meta}},
						category => $type,
						structure => $structure,
						link => $link,
					};
					$root = $link;
				}

				# The links themselves
				foreach my $category (keys %{$element->{$type}{links}}) {
					my $link = URI->new($element->{$type}{links}{$category})->abs($root);
					push @{$links{$root}}, {
						structure => $structure,
						category => $category,
						type => $type,
						link => $link,
					};
				}
			}

		} else {
			die 'No links';
		}
	}

	# Canonicalize the links
	$_->{link} = URI->new($_->{link})->abs($root) foreach @{$links{$root}};

	my @matches = grep {
		my $this_link = $_;
		# Filter out those, who lack any of our keys or
		# hold a different value for it.
		not map { not exists $link->{$_}
			or not exists $this_link->{$_}
			or $link->{$_} ne $this_link->{$_}
			? 1 : () } keys %$link
	} @{$links{$root}};

	# Fully resolved
	return @matches unless @path;

	die 'Nonexistent component in path' unless @matches;
	die 'Ambigious path' unless scalar @matches == 1;

	# Traverse further
	return $self->get_canonical_links($matches[0]->{link}, @path);
}

# This is a 'normalized' version, for convenience and compatibility
sub get_links {
	my $self = shift;
	my $root = (ref $_[0] and ref $_[0] ne 'HASH') ? shift : '';

	# Canonicalize URIs
	$root = URI->new($root)->abs($self->{agent}{root});

	# And decanonicalize, ommiting the scheme and authority part if possible
	my @links = $self->get_canonical_links($root, @_);
	$_->{link} = $_->{link}->rel($root)->authority
		?  $_->{link} : URI->new($_->{link}->path) foreach @links;

	return @links;
}

=item B<links> PATH

Traverse the links in resource hierarchy following given PATH,
starting from API root (L</gdc> by default).

PATH is an array of dictionaries, where each key-value pair
matches properties of a link. If a plain string is specified,
it is considered to be a match against B<category> property:

  $gdc->links('md', { 'category' => 'projects' });

The above call returns a list of all projects, with links to
their metadata resources.

=cut

sub links {
	my @links = get_links @_;
	return @links if @links;
	%links = ();
	return get_links @_;
}

=item B<get_uri> PATH

Follows the same samentics as B<links>() call, but returns an
URI of the first matching resource instead of complete link
structure.

=cut

sub get_uri {
	[links @_]->[0]{link};
}

=item B<login> EMAIL PASSWORD

Obtain a SST (login token).

=cut

sub login {
	my $self = shift;
	my ($login, $password) = @_;

	$self->{login} = $self->{agent}->post($self->get_uri('login'),
		{postUserLogin => {
			login => $login,
			password => $password,
			verify_level => 2,
			remember => 0}});

	$self->{agent}{GDCAuthSST} = $self->{login}{userLogin}{token};
}

=item B<logout>

Make server invalidate the client session and drop
credential tokens.

Is called upon destruction of the GoodData client instance.

=cut

sub logout {
	my $self = shift;

	die 'Not logged in' unless defined $self->{login};

	$self->{agent}->delete($self->{login}{userLogin}{state},
		'X-GDC-AuthSST' => $self->{agent}{GDCAuthSST});

	# Maybe this needs to go to the agent itself?
	delete $self->{agent}{GDCAuthSST};
	delete $self->{agent}{GDCAuthTT};
	$self->{agent}->default_header('X-GDCAuthTT' => undef);

	$self->{login} = undef;
}

=item B<change_passwd> OLD NEW

Change user password given the old and new password.

=cut

sub change_passwd {
	my $self = shift;
	my $old_passwd = shift or die 'No old password given';
	my $new_passwd = shift or die 'No new password given';

	die 'Not logged in' unless defined $self->{login};

	my $profile = $self->{agent}->get($self->{login}{userLogin}{profile});
	my $new_profile = {
		'accountSetting' => {
			'old_password' => $old_passwd,
			'password' => $new_passwd,
			'verifyPassword' => $new_passwd,
			'firstName' => $profile->{accountSetting}->{firstName},
			'lastName' => $profile->{accountSetting}->{lastName}
		}
	};

	$self->{agent}->put($self->{login}{userLogin}{profile}, $new_profile);
}

=item B<projects>

Return array of links to project resources on metadata server.

=cut

sub projects {
	my $self = shift;

	die 'Not logged in' unless $self->{login};
	$self->get_links(URI->new($self->{login}{userLogin}{profile}),
		qw/projects project/);
}

=item B<delete_project> IDENTIFIER

Delete a project given its identifier.

=cut

sub delete_project {
	my ($self, $project) = @_;

	# Instead of directly DELETE-ing the URI gotten, we check
	# the existence of a project with such link, as a sanity check
	my $uri = $self->get_uri(URI->new($project),
		{ category => 'self', type => 'project' }) # Validate it's a project
		or die "No such project: $project";
	$self->{agent}->delete($uri);
}

=item B<create_project> TITLE SUMMARY TEMPLATE DRIVER TOKEN

Create a project given its title and optionally summary, project template,
db engine driver and authorization token
return its identifier.

The list of valid project templates is available from the template server:
L<https://secure.gooddata.com/projectTemplates/>.

Valid db engine drivers are 'Pg' (default) and 'vertica'.

=cut

sub create_project {
	my $self = shift;
	my $title = shift or die 'No title given';
	my $summary = shift || '';
	my $template = shift;
	my $driver= shift;
	my $token = shift;
	my $environment = shift;

	# The redirect magic does not work for POSTs and we can't really
	# handle 401s until the API provides reason for them...
	$self->{agent}->get($self->get_uri('token'));

	return $self->{agent}->post($self->get_uri('projects'), {
		project => {
			content => {
				# No hook to override this; use web UI
				guidedNavigation => 1,
				($driver ? (driver => $driver) : ()),
				($environment ? (environment => $environment) : ()),
				($token ? (authorizationToken => $token) : ())
			},
			meta => {
				summary => $summary,
				title => $title,
				($template ? (projectTemplate => $template) : ()),
			}
	}})->{uri};
}

=item B<create_user> DOMAIN EMAIL LOGIN PASSWORD FIRST_NAME LAST_NAME PHONE COMPANY SSO_PROVIDER

Create a user given its email, login, password, first name, surname, phone and optionally company,
sso provider in domain.

Returns user identifier (URI).

=cut

sub create_user {
	my $self = shift;
	my $domain_uri = shift || die "No domain specified";
	my $email = shift || die "Email must be specified";
	my $login = shift || $email;
	my $passwd = shift;
	my $firstname = shift;
	my $lastname = shift;
	my $phone = shift;
	my $company = shift || '';
	my $sso_provider = shift;

	return $self->{agent}->post($domain_uri."/users", { #TODO links does not exists in REST API
		accountSetting => {
			login => $login,
			email => $email,
			password => $passwd,
			verifyPassword => $passwd,
			firstName => $firstname,
			lastName => $lastname,
			phoneNumber => $phone,
			companyName => $company,
			($sso_provider ? (ssoProvider => $sso_provider) : ()),
	}})->{uri};
}

=item B<get_roles> PROJECT

Gets project roles.

Return array of project roles.

=cut

sub get_roles {
	my ($self, $project) = @_;

	return $self->{agent}->get(
		$self->get_uri(URI->new($project), 'roles'))->{projectRoles}{roles};
}
=item B<reports> PROJECT

Return array of links to repoort resources on metadata server.

=cut

sub reports {
	my ($self, $project) = @_;

	die 'Not logged in' unless $self->{login};
	$self->get_links(
		URI->new($project),
		{ category => 'self', type => 'project' }, # Validate it's a project
		qw/metadata query reports/, {},
	);
}

=item B<compute_report> REPORT

Trigger a report computation and return the URI of the result resource.

=cut

sub compute_report {
	my ($self, $report) = @_;

	return $self->{agent}->post(
		$self->get_uri(qw/xtab xtab-executor3/),
		{ report_req => { report => ''.$report }},
	);
}

=item B<export_report> REPORT FORMAT

Submit an exporter task for a computed report (see B<compute_report>),
wait for completion and return raw data in desired format.

=cut

sub export_report {
	my ($self, $report, $format) = @_;

	# Compute the report
	my $result = $self->{agent}->post(
		$self->get_uri(qw/report-exporter exporter-executor/),
		{
			result_req => {
				format => $format,
				result => $self->compute_report($report),
			},
		},
	);

	# This is for new release, where location is finally set correctly;
	$result = $result->{uri} if ref $result eq 'HASH';

	# Trigger the export
	my $exported = $self->poll(
		sub { $self->{agent}->get($result) },
		sub { $_[0] and exists $_[0]->{raw} and $_[0]->{raw} ne 'null' }
	) or die 'Timed out waiting for report to export';

	# Follow the link
	$exported = $self->{agent}->get($exported->{uri}) if exists $exported->{uri};

	# Gotten the correctly coded result?
	my $wanted = $format;
	my %compat = (
		png => 'image/png',
		pdf => 'application/pdf',
		xls => 'application/vnd.ms-excel',
		csv => 'text/csv',
	);
	$wanted = $compat{$wanted} if exists $compat{$wanted};
	return $exported->{raw} if $exported->{type} =~ /^$wanted/;

	die 'Wrong type of content returned';
}

=item B<ldm_picture> PROJECT

Return picture of Logical Data Model (LDM) in PNG format.

=cut

sub ldm_picture {
	my $self = shift;
	my $project = shift;

	my $model = $self->{agent}->get(
		$self->{agent}->get(
			$self->get_uri(
				URI->new($project),
				{ category => 'ldm' },
			),
		)->{uri},
	);
	die 'Expected PNG image' unless $model->{type} eq 'image/png';

	return $model->{raw};
}

=item B<ldm_manage> PROJECT MAQL

Execute MAQL statement for a project.

=cut

sub ldm_manage {
	my ($self, $project, $maql) = @_;

	$maql = "# WWW::GoodData MAQL execution\n$maql";
	chomp $maql;

	$self->{agent}->post(
		$self->get_uri(
			URI->new($project),
			qw/metadata ldm ldm-manage/,
		),
		{ manage => { maql => $maql } },
	);
}

=item B<upload> PROJECT MANIFEST

Upload and integrate a new data load via Single Loading Interface (SLI).

=cut

sub upload {
	my ($self, $project, $file) = @_;

	# Parse the manifest
	my $upload_info = decode_json(slurp_file($file));
	die "$file: not a SLI manifest"
		unless $upload_info->{dataSetSLIManifest};

	# Construct unique URI in staging area to upload to
	my $uploads = URI->new($self->get_uri('uploads'));
	$uploads->path_segments(
		$uploads->path_segments,
		$upload_info->{dataSetSLIManifest}{dataSet}.'-'.time,
	);
	$self->{agent}->request(HTTP::Request->new(MKCOL => $uploads));

	# Upload the manifest
	my $manifest = $uploads->clone;
	$manifest->path_segments($manifest->path_segments, 'upload_info.json');
	$self->{agent}->request(
		HTTP::Request->new(
			PUT => $manifest,
			['Content-Type' => 'application/json'],
			encode_json($upload_info),
		),
	);

	# Upload CSV
	my $csv = $uploads->clone;
	$csv->path_segments($csv->path_segments, $upload_info->{dataSetSLIManifest}{file});
	$self->{agent}->request(
		HTTP::Request->new(
			PUT => $csv,
			['Content-Type' => 'application/csv'],
			(slurp_file($upload_info->{dataSetSLIManifest}{file})
				|| die 'No CSV file specified in SLI manifest')
		),
	);

	# Trigger the integration
	my $task_url = $self->{agent}->post(
		$self->get_uri(
			URI->new($project),

			# Validate it's a project
			{
				category => 'self',
				type => 'project',
			},
			qw/metadata etl pull2/,
		), {
			pullIntegration => [$uploads->path_segments]->[-1],
		}
	)->{pull2Task}{links}{poll};

	return $self->poll($task_url);
}

=item B<uploads> FILE_PATH FILE UPLOAD_PATH

Uploads to WebDAV service.

=cut

sub uploads {
	my ($self, $file_path, $upload_path) = @_;
	if (! -r $file_path) {
		die "Cannot read '$file_path' file for upload.";
	}
	my $uploads = URI->new($self->get_uri('uploads'));
	my @collections = split m/\//ms, $upload_path;
	my $file = pop @collections;
	foreach my $collection (@collections) {
		$uploads->path_segments($uploads->path_segments, $collection);
		if (! $self->_collection_exists($uploads)) {
			$self->{agent}->request(HTTP::Request->new(MKCOL => $uploads));
		}
	}
	$uploads->path_segments($uploads->path_segments, $file);
	$self->{agent}->request(HTTP::Request->new(
		PUT => $uploads,
		undef,
		slurp_file($file_path),
	));
	return $uploads;
}

sub _collection_exists {
	my ($self, $uploads) = @_;
	my $wd_check_col_xml = <<'END';
<?xml version="1.0"?>
<propfind xmlns="DAV:">
   <prop>
      <resourcetype />
   </prop>
</propfind>
END

	# XXX Hack because WWW::GoodData::Agent changes LWP::UserAgent behavior.
	eval {
		$self->{agent}->request(HTTP::Request->new(
			PROPFIND => $uploads,
			undef,
			$wd_check_col_xml,
		));
	};
	if ($EVAL_ERROR =~ '^404') {
		return 0;
	} else {
		return 1;
	}
}

=item B<poll> TASK_URL

Periodically checks status of task page and check 'wTaskStatus':'status' for 'RUNNING'.

Returns undef is status is OK.

=cut

sub poll {
        my ($self, $task_url) = @_;

	# Wait for the task to enter a stable state
	my $result = $self->_poll(
		sub { $self->{agent}->get($task_url) },
		sub { shift->{wTaskStatus}{status} !~ /^(RUNNING)$/ }
	) or die 'Timed out waiting for integration to finish';

	return if $result->{wTaskStatus}{status} eq 'OK';

	die
		'Upload finished with '.$result->{wTaskStatus}{status}." status and message:\n".
		$result->{wTaskStatus}{messages}[0]{error}{message}."\nand parameters:\n".
		join("; ", @{$result->{wTaskStatus}{messages}[0]{error}{parameters}})."\n"
	;
}

sub _poll {
        my ($self, $body, $cond) = @_;

        my $retries = $self->{retries};
        while ($retries--) {
                my $ret = $body->();
                return $ret if $cond->($ret);
                sleep 1;
        }

        return undef;
}

=item B<create_object_with_expression> PROJECT URI TYPE TITLE SUMMARY EXPRESSION

Create a new metadata object of type TYPE with EXPRESSION as the only content.

=cut

sub create_object_with_expression {
	my $self = shift;
	my $project = shift;
	my $uri = shift;
	my $type = shift or die 'No type given';
	my $title = shift or die 'No title given';
	my $summary = shift || '';
	my $expression = shift or die 'No expression given';

	if (defined $uri) {
		$uri = URI->new($uri);
	} else {
		$uri = $self->get_uri(URI->new($project), qw/metadata obj/);
	}

	return $self->{agent}->post(
		$uri,
		{
			$type => {
				content => {
					expression => $expression
				},
				meta => {
					summary => $summary,
					title => $title,
				}
			},
		},
	)->{uri};
}

=item B<create_report_definition> PROJECT URI TITLE SUMMARY METRICS DIM FILTERS

Create a new reportDefinition in metadata.

=cut

sub create_report_definition {
	my $self = shift;
	my $project = shift;
	my $uri = shift;
	my $title = shift or die 'No title given';
	my $summary = shift || '';
	my $metrics = shift || [];
	my $dim = shift || [];
	my $filters = shift || [];

	if (defined $uri) {
		$uri = URI->new($uri);
	} else {
		$uri = $self->get_uri(URI->new($project), qw/metadata obj/);
	}

	return $self->{agent}->post(
		$uri,
		{
			reportDefinition => {
				content => {
					filters => [ map +{ expression => $_ }, @$filters ],
					grid => {
						columns => [ "metricGroup" ],
						metrics => [ map +{ alias => '', uri => $_ }, @$metrics ],
						rows => [ map +{ attribute => { alias => '', uri => $_,
							totals => [[]] } }, @$dim ],
						sort => {
							columns => [],
							rows => [],
						},
						columnWidths => [],
					},
					format => "grid",
				},
				meta => {
					summary => $summary,
					title => $title,
				},
			},
		},
	)->{uri};
}

=item B<DESTROY>

Log out the session with B<logout> unless not logged in.

=cut

sub DESTROY {
	my $self = shift;
	return unless $self->{auto_destroy};
	$self->logout if $self->{login};
}

=item B<dd_pull> PROJECT ZIP_FILE DATASETS

Integrate a custom date dimension load via Single Loading Interface (SLI).

=cut

sub dd_pull {
	my ($self, $project, $zip_file, $datasets) = @_;

	# Trigger the integration
	my $task_url = $self->{agent}->post(
		$self->get_uri(
			URI->new($project),

			# Validate it's a project
			{
				category => 'self',
				type => 'project',
			},
			qw/metadata datedimension pull/
		), {
			dateIntegration => {
				file => $zip_file,
				datasets => $datasets,
			},
		},
	)->{asyncTask}{link}{poll};

	return $self->poll($task_url);
}

sub slurp_file {
        my $file = shift;
        open (my $fh, '<', $file) or die "$file: $!";
        return join '', <$fh>;
}

sub _hide_password {
	my ($self, $content) = @_;
	return $content if $content !~ 'postUserLogin';
	my $content_json = decode_json($content);
	return $content unless exists $content_json->{postUserLogin}->{password};
	$content_json->{postUserLogin}->{password} = '***';
	return encode_json($content_json);
}

=back

=head1 SEE ALSO

=over

=item *

L<http://developer.gooddata.com/api/> -- API documentation

=item *

L<https://secure.gooddata.com/gdc/> -- Browsable GoodData API

=item *

L<WWW::GoodData::Agent> -- GoodData API-aware user agent

=back

=head1 CAVEATS

=over 8

=item * uploads()

Don't check, when collection exists.

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2017 Michal Josef Spacek

Copyright 2011, 2012, 2013 Lubomir Rintel

Copyright 2012, 2013 Adam Stulpa, Jan Orel, Tomas Janousek

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=head1 AUTHORS

Tomas Janousek C<tomi@nomi.cz>

Jan Orel C<jan.orel@gooddata.com>

Lubomir Rintel C<lkundrak@v3.sk>

Michal Josef Spacek C<michal.spacek@gooddata.com>

Adam Stulpa C<adam.stulpa@gooddata.com>

=cut

1;
